src/main/java/com/google/devtools/build/lib/bazel/rules/python/BazelPythonSemantics.java

```diff
@@ -237,7 +237,7 @@ public void createExecutable(
         PathFragment shExecutable = ShToolchain.getPathOrError(ruleContext);
         // TODO(#8685): Remove this special-case handling as part of making the proper shebang a
         // property of the Python toolchain configuration.
-        String pythonExecutableName = OS.getCurrent() == OS.OPENBSD ? "python3" : "python";
+        String pythonExecutableName = "python3";
```

Comment:
Won't this break anyone who is still using Python3?

cc: @brandjon

Reply:
@aiuto assuming you mean Python 2 then, yes. If a user has *only* Python 2 installed and not Python 3 then this would fail for them. However, I'm not sure how many users, even those who use Python 2 as default, don't have Python 3 installed at all. (Or couldn't easily install it) Debian ships it even in our oldoldstable (which is really obsolete).

Reply:
FYI macOS before Catalina only had python2 installed by default, so there is no global /usr/bin/python3, _some_ users might have that installed through homebrew at /usr/local/bin/python3, but it's not a requirement today and arguably relying on brewed python is an issue since users could have any sub-version there depending on when they installed it

Reply:
We can't require that Python3 be installed on all target platforms where Python 2 targets run.

The proper solution is probably to review #11434 and make the autodetecting toolchain for Python 3 emit a `python3` shebang. The only case I can think of where this would be a breaking change is where there's no `python3` command in the environment, but there is a `python` command that resolve to a Python 3 interpreter. In which case I'm tempted to just say add an explicit toolchain to your build.

Reply:
This code path only affects the disabled-by-default "launchable Python zips" in the deprecated native Python rules of Bazel. For example, if you build something like this:

```
philwo@philwo-macbookpro2:src/pytest $ cat BUILD 
py_binary(
    name = "test",
    srcs = ["test.py"],
)

philwo@philwo-macbookpro2:src/pytest $ cat test.py 
print("Hello world!")

philwo@philwo-macbookpro2:src/pytest $ bazel build --build_python_zip :test
Target //:test up-to-date:
  bazel-bin/test
  bazel-bin/test.zip
```

Then the `bazel-bin/test` would be a ZIP file with the shebang prepended, so you can just run it as an executable, but only if you have a `python3` executable in your path. Without this change you can only run it if you have a `python` executable in your path.

The `python` executable on the path means "Python 2" on almost all distros (only Arch Linux was bold enough to migrate that to Python 3 :)). So currently we require everyone to install an outdated, EOL'd version of Python to run these binaries, even if their own code is Python 3, as it should be.

With the change, we no longer require Python 2.x installed, but Python 3. IMHO this is fine, especially considering https://python3statement.org/.

Reply:
@brandjon's suggestion is obviously the cleanest one, but IMHO (just speaking for myself here) any time invested into supporting Python 2 is wasted at this point.

@keith You have a good point there, but looking forward, Apple also said that [you should not rely](https://developer.apple.com/documentation/macos_release_notes/macos_catalina_10_15_release_notes) on any Python version being installed by default on macOS as they might remove them in the future. So this will all become moot in a few months if they actually go ahead and remove pre-installed Python or make it unusable.

Reply:
I think our comments crossed, so interesting that we have opposite conclusions. :)

> deprecated native Python rules of Bazel

Deprecated in the sense that we want people to use stubs instead? Sure, but the rules themselves aren't deprecated, they just haven't been migrated to Starlark yet.

The Bazel logo isn't on the python3statement.org; we still support building code that uses Python 2, and we'll need to make a conscious decision to drop such support before we can change the stub in that way. In the meantime I *think* allowing configuring the shebang should suffice.

You may be right about wasted effort, but I suspect the shebang thing is useful even outside of just selecting the interpreter version.

Reply:
> Deprecated in the sense that we want people to use stubs instead?

Sorry, I mixed this up - I thought our `rules_python` were already a new implementation independent of the native rules.

> The Bazel logo isn't on the python3statement.org

Yes, but we made an [announcement for Bazel 2.0](https://groups.google.com/forum/#!msg/bazel-discuss/isTkOF-av_A/iV3UIfB0AQAJ) referencing that page and saying that we commit to it. Admittedly we wrote there that we'd still support building Python 2.x code. :/

What would it take to make that conscious decision to drop Python 2.x support? I would really like to see us do so, of course following all required steps for a change like that.

Reply:
> You may be right about wasted effort, but I suspect the shebang thing is useful even outside of just selecting the interpreter version.

Yes, that sounds like a nice way. We could let it default to the current `python` one and introduce an incompatible flag that flips it to `python3`, WDYT?

Reply:
> Admittedly we wrote there that we'd still support building Python 2.x code. :/

Which was a very intentional policy point, not an oversight. Google isn't even free of Python 2 code, so we can't expect all our users to be. (Of course, we're talking about the ability to build and run Python 2 targets, which isn't the same thing as requiring Python 3 to also be present.)

What would dropping Python 2 support look like in your ideal vision? Would we no longer have a `python_version` attribute that chose between PY2 and PY3? To drop support of the whole concept of running PY2 code from the Python rules, I'd want to see virtually no users having any PY2 code in their builds. Given that only last year we changed the default and it broke people, I think we have some time before that happens. Perhaps enough that we'll first migrate the rules entirely to Starlark and it'll be a different calculus (e.g. if the version selection system gets more granularity).

> We could let it default to the current python one and introduce an incompatible flag that flips it to python3, WDYT?

If people are not setting up an explicit Python toolchain in their build, then they're using the autodetecting toolchain, which will look for Python 2 or Python 3 depending on the configuration of the target. It's easy for these toolchains to specify custom shebangs that use `python` and `python3` respectively. (Of course, the `python` command on Arch will yield a Python 3 interpreter, but the stub script doesn't care about the version so long as it's at least present.)

If people are setting up a custom toolchain, they can modify their toolchain definition to explicitly specify a shebang.

For the default value, of a toolchain that is not the autodetecting one, we can probably use an incompatible flag to nudge people along. But this would be for convenience; it doesn't block anything.

Reply:
Ah, just looked at #11434 and I can now understand the %shebang% functionality better. Is there a consensus here to change the [default](https://github.com/beasleyr-vmw/bazel/blob/6e433a29e3e814dc487df69882e911421b0b2a40/src/main/java/com/google/devtools/build/lib/skylarkbuildapi/python/PyRuntimeInfoApi.java#L46) to Python 3 in that mechanism as well? I would personally recommend that but I'll hold off on adding that to this PR lacking said consensus.

Reply:
@olekw Yes, let's do that and change the default there to python3.

Reply:
I'll merge this and send it for internal review, then we can do a follow-up change on that line.